packages/workbox-background-sync/lib/Queue.mjs

```diff
@@ -0,0 +1,297 @@
+ * // Manually detecting failed requests and added them to the queue.
+ * const bgQueue = new workbox.backgroundSync.Queue('myQueue');
+ * self.addEventListener('fetch', function(event) {
+ *   if (!event.request.url.startsWith('https://example.com')) return;
+ *
+ *   const clone = event.request.clone();
+ *   event.respondWith(fetch(event.request).catch((err) => {
+ *     bgQueue.addRequest(clone);
+ *     throw err;
+ *   }));
+ * });
+ *
+ * @memberof module:workbox-background-sync
+ */
+export default class Queue {
+  /**
+   * Creates an instance of Queue with the given options
+   *
+   * @param {string} name The unique name for this queue. This name must be
+   *     unique as it's used to register sync events and store requests
+   *     in IndexedDB specific to this instance. An error will be thrown if
+   *     a duplicate name is detected.
+   * @param {Object} [param2]
+   * @param {number} [param2.maxRetentionTime = 7 days] The amount of time (in
+   *     ms) a request may be retried. After this amount of time has passed,
+   *     the request will be deleted and not retried.
+   * @param {Object} [param2.callbacks] Callbacks to observe the lifecycle of
+   *     queued requests. Use these to respond to or modify the requests
+   *     during the replay process.
+   * @param {function(StorableRequest):undefined}
+   *     [param2.callbacks.requestWillEnqueue]
+   *     Invoked immediately before the request is stored to IndexedDB. Use
+   *     this callback to modify request data at store time.
+   * @param {function(StorableRequest):undefined}
+   *     [param2.callbacks.requestWillReplay]
+   *     Invoked immediately before the request is re-fetched. Use this
+   *     callback to modify request data at fetch time.
+   * @param {function(StorableRequest):undefined}
+   *     [param2.callbacks.requestDidReplay]
+   *     Invoked immediately after the request has successfully re-fetched.
+   * @param {function(StorableRequest):undefined}
+   *     [param2.callbacks.replayDidFail]
+   *     Invoked if the replay attempt failed.
+   * @param {function(Array<StorableRequest>):undefined}
+   *     [param2.callbacks.allRequestsDidReplay]
+   *     Invoked after all requests in the queue have successfully replayed.
+   */
+  constructor(name, {
+    maxRetentionTime = MAX_RETENTION_TIME,
+    callbacks = {},
+  } = {}) {
+    // Ensure the store name is not already being used
+    if (names.has(name)) {
+      throw new WorkboxError('duplicate-queue-name', {name});
+    } else {
+      names.add(name);
+    }
+
+    this._name = name;
+    this._callbacks = callbacks;
+    this._maxRetentionTime = maxRetentionTime;
+
+    this._addSyncListener();
+  }
+
+  /**
+   * Returns an object containing the `fetchDidFail` lifecycle method. This
+   * object can be used as a RequestWrapper plugin to automatically add
+   * failed requests to the background sync queue to be retried later.
+   *
+   * @return {Object}
+   */
+  createPlugin() {
```

Comment:
I added this because it was in the original API design. Here's the original code example (modified to use `createPlugin()`):

```js
const queue = new workbox.backgroundSync.Queue('myQueue');

const requestWrapper = new workbox.runtimeCaching.RequestWrapper({
  plugins: [queue.createPlugin()],
});

const route = new workbox.routing.RegExpRoute({
  regExp: new RegExp('^https://jsonplaceholder.typicode.com'),
  handler: new workbox.runtimeCaching.NetworkOnly({requestWrapper}),
});

const router = new workbox.routing.Router();
router.registerRoute({route});
```

Reply:
I'd drop this in favor of exposing a plugin class like the other plugins. While I understand the convenience of this, nothing else follows this pattern.

Reply:
I don't like the v2 way (where the plugin class just extends the `Queue` class), but if you want it to be a class instead of an object, it could be like this:

```js
class QueuePlugin {
  constructor(queue) {
    this._queue = queue
  }
  fetchDidFail({request}) {
    this._queue.addRequest(request);
  }
}
```

And you'd use it like this:

```js
new workbox.runtimeCaching.NetworkOnly({
  plugins: new QueuePlugin(queue),
});
```

vs. how I have it currently:

```js
new workbox.runtimeCaching.NetworkOnly({
  plugins: queue.createPlugin(),
});
```

* * *

Alternatively, we could just do away with this feature entirely, since it's really not that hard to do this:

```js
new workbox.runtimeCaching.NetworkOnly({
  plugins: {
    fetchDidFail: ({request}) {
      queue.addRequest(request);
    }
  }
});
```

WDYT?

Reply:
I'd vote for the first option.

If we go with explicitly needing to know the lifecycle callback names to use this behavior, then the precedent would be to use that approach with all the other plugins as well.

That isn't a great user experience for developers who care about specifying behaviors, but don't care about the underlying lifecycle callbacks. It's also an issue for the other behaviors that require multiple callbacks (I think cache expiration does?), since it's even harder to properly configure that.

Reply:
SGTM. Done.